### PR TITLE
disttask: refine scheduler error handling

### PR DESCRIPTION
**Pull request by ywqzzy**
### What problem does this PR solve?

Issue Number: ref #46258

Problem Summary:
When tidb gracefully shutdown, dist task failed.
When network partition with tikv, dist task failed due to retryable error.
### What is changed and how it works?
1. use backoff for functions related to tikv in scheduler.
2. check if error retryable, if retryable, don't mark subtask as failed.
3. use cancelCauseFunc to cancel subtasks. Then when tidb gracefully shutdown, subtask will not be marked as canceled.
```
func (s *BaseScheduler) markTaskCancelOrFailed(ctx context.Context, subtask *proto.Subtask) bool {
	if err := s.getError(); err != nil {
		if ctx.Err() != nil && context.Cause(ctx).Error() == "cancel subtasks" {
			logutil.Logger(s.logCtx).Warn("subtask canceled", zap.Error(err))
			s.updateSubtaskStateAndError(subtask, proto.TaskStateCanceled, nil)
		} else if common.IsRetryableError(err) {
			logutil.Logger(s.logCtx).Warn("met retryable error", zap.Error(err))
		} else if errors.Cause(err) != context.Canceled {
			logutil.Logger(s.logCtx).Warn("subtask failed", zap.Error(err))
			s.updateSubtaskStateAndError(subtask, proto.TaskStateFailed, err)
		} else {
			logutil.Logger(s.logCtx).Warn("met context canceled for gracefully shutdown", zap.Error(err))
		}
		s.markErrorHandled()
		return true
	}
	return false
}
```
We use the above function to persist error to subtasks.
### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   
image
HA test tidb to all network partition passed.
Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: disttask/framework/scheduler/manager.go [hunk 1]
```diff
@@ -32,8 +32,7 @@ import (
 )
 
 var (
-	schedulerPoolSize       int32 = 4
-	subtaskExecutorPoolSize int32 = 10
```
Comment (ywqzzy) [inline disttask/framework/scheduler/manager.go:36]: useless

File: disttask/framework/scheduler/scheduler.go [hunk 2]
```diff
@@ -236,23 +256,32 @@ func (s *BaseScheduler) run(ctx context.Context, task *proto.Task) error {
...
+
+	finished := s.markTaskCancelOrFailed(ctx, subtask)
+	if finished {
 		return
 	}
+
 	if ctx.Err() != nil {
 		s.onError(ctx.Err())
-		return
+		finished := s.markTaskCancelOrFailed(ctx, subtask)
```
Comment (D3Hunter) [inline disttask/framework/scheduler/scheduler.go:281]: hard to read to call those function multiple times😅

File: disttask/framework/scheduler/scheduler.go [hunk 3]
```diff
@@ -168,6 +186,10 @@ func (s *BaseScheduler) run(ctx context.Context, task *proto.Task) error {
 		proto.TaskStatePending, proto.TaskStateRunning)
 	if err != nil {
 		s.onError(err)
```
Comment (wjhuang2016) [inline disttask/framework/scheduler/scheduler.go:188]: Do we need to record the error if it is retryable?

File: disttask/framework/scheduler/manager.go [hunk 4]
```diff
@@ -67,7 +66,7 @@ type Manager struct {
 		sync.RWMutex
 		// taskID -> cancelFunc.
 		// cancelFunc is used to fast cancel the scheduler.Run.
```
[the remark this answers is not in the document]
Comment (ywqzzy) [reply disttask/framework/scheduler/manager.go:68]: Yes, will fix it

**Review by Benjamin2037 — APPROVED** [review]
LGTM

File: disttask/framework/scheduler/scheduler.go [hunk 5]
```diff
@@ -473,38 +504,102 @@ func (s *BaseScheduler) resetError() {
...
 	metrics.EndDistTaskSubTask(subtask)
-	if err := s.taskTable.FinishSubtask(subtask.ID, subtask.Meta); err != nil {
-		s.onError(err)
-	}
+	s.finishSubtask(ctx, subtask)
 	subtask.State = proto.TaskStateSucceed
 	metrics.IncDistTaskSubTaskCnt(subtask)
 }
+
+func (s *BaseScheduler) markTaskCancelOrFailed(ctx context.Context, subtask *proto.Subtask) bool {
```
Comment (D3Hunter) [inline disttask/framework/scheduler/scheduler.go:577]:
```suggestion
func (s *BaseScheduler) markSubtaskCancelOrFailed(ctx context.Context, subtask *proto.Subtask) bool {
```

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

**Pull request merged**